source/common/quic/platform/quiche_flags_impl.cc

```diff
@@ -9,19 +9,89 @@
 #include "absl/strings/numbers.h"
 #include "quiche_platform_impl/quiche_flags_impl.h"
 
+namespace {
+
+// Envoy uses different default values for some QUICHE flags. The following methods
+// ensure that the absl::Flag objects are created with the correct values for
+// these flags. This ensures that the absl::FlagSaver finds the correct values
+// and avoid a race condition between the dynamic initialization of these flags
+// and the FlagSaver in tests.
+template <typename T> constexpr T maybeOverride(absl::string_view /*name*/, T val) { return val; }
+
+template <> constexpr bool maybeOverride<bool>(absl::string_view name, bool val) {
+  if (name == "quic_reloadable_flag_quic_disable_version_draft_29") {
+    // Envoy only supports RFC-v1 in the long term, so disable IETF draft 29 implementation by
+    // default.
+    return true;
+  }
+  if (name == "quic_reloadable_flag_quic_default_to_bbr") {
+    // This flag enables BBR, otherwise QUIC will use Cubic which is less performant.
+    return true;
+  }
+  if (name == "quic_header_size_limit_includes_overhead") {
+    // Do not include 32-byte per-entry overhead while counting header size.
+    return false;
+  }
+
+  return val;
+}
+
+template <> constexpr int32_t maybeOverride<int32_t>(absl::string_view name, int32_t val) {
+  if (name == "quic_buffered_data_threshold") {
+    // Set send buffer twice of max flow control window to ensure that stream send
+    // buffer always takes all the data.
+    // The max amount of data buffered is the per-stream high watermark + the max
+    // flow control window of upstream. The per-stream high watermark should be
+    // smaller than max flow control window to make sure upper stream can be flow
+    // control blocked early enough not to send more than the threshold allows.
+    // TODO(#8826) Ideally we should use the negotiated value from upstream which is not accessible
+    // for now. 512MB is way to large, but the actual bytes buffered should be bound by the
+    // negotiated upstream flow control window.
+    return 2 * ::Envoy::Http2::Utility::OptionsLimits::DEFAULT_INITIAL_STREAM_WINDOW_SIZE; // 512MB
+  }
+  return val;
+}
+
+} // namespace
+
+// Flag definitions
+#define QUIC_FLAG(flag, value) ABSL_FLAG(bool, flag, maybeOverride(#flag, value), "");
+#include "quiche/quic/core/quic_flags_list.h"
+QUIC_FLAG(quic_reloadable_flag_spdy_testonly_default_false, false)
+QUIC_FLAG(quic_reloadable_flag_spdy_testonly_default_true, true)
+QUIC_FLAG(quic_restart_flag_spdy_testonly_default_false, false)
+QUIC_FLAG(quic_restart_flag_spdy_testonly_default_true, true)
+QUIC_FLAG(quic_reloadable_flag_http2_testonly_default_false, false)
+QUIC_FLAG(quic_reloadable_flag_http2_testonly_default_true, true)
+QUIC_FLAG(quic_restart_flag_http2_testonly_default_false, false)
+QUIC_FLAG(quic_restart_flag_http2_testonly_default_true, true)
+#undef QUIC_FLAG
+
+#define QUIC_PROTOCOL_FLAG(type, flag, value, help)                                                \
```

Comment:
For protocol flags, we may have extra arg `external_value` in these macros. So I think the QUIC_PROTOCOL_FLAG_MACRO_CHOOSER handling is still needed.

Reply:
It doesn't look like we actually do this any longer, since none of the flags have such a value. But I guess it could come back later. Done.